Make `Simulator` class and exportation methods

qian-chu opened the pull request:
Tutorial also updated (code)

To-do list:

- [x] Add export to eeglab (using eeglabio), which requires specification of file paths
- [x] Further modify docstr and tutorial to make them compatible with the current code

qian-chu committed:
refactor the package, convert the main simulate function to a class and exportation to other formats as methods

AlexLepauvre committed:
Add function to create events and events id

The generate mne object can be populated with the events information for later extraction of the data

AlexLepauvre committed:
Add option to export to eeg lab

AlexLepauvre committed:
Update doc

AlexLepauvre commented:
@qian-chu, I have added a method to generate events and events id to add to the mne epochs, based directly on the parameters of the simulator class, with additional handy optional inputs. That was necessary for the export to eeglab, which otherwise complains about not having any events info. 
I have modified the jupyter notebook to add the info there. Let me know if you think that's enough. 

I will now add the following:

- [x] Add a kernel option to convolve the box car with
- [x] Add a between subject noise option, as currently there is no such option

Then we can decide whether to add pytest scripts.

AlexLepauvre committed:
Add temporal kernel option

To specify the temporal dynamics of the simulated signal

AlexLepauvre committed:
Add tutorial for kernel specification

Showcase how to use a temporal kernel. Example: gamma kernel

AlexLepauvre commented:
I've added the option of a simple temporal kernel, with a tutorial notebook. Let me know what you think

AlexLepauvre committed:
Rename first tutorial

Clearer name for later extension of tutorials

AlexLepauvre committed:
Add intersubject noise term

AlexLepauvre committed:
Add corresponding documentation

AlexLepauvre commented:
I've added the between subject noise parameter and added a corresponding tutorial. One last thing I want to add (and then we are done I promise) is an example of a sample size calculation based on our pipeline

qian-chu commented:
Sounds good, I have some further pitches on top:

- [x] Terminology of "mode" and "space": I found it a bit confusing. Do you think calling it "channel" will be more intuitive? This way we also conform to MNE epoch convention too.
- [x] Now you are using [`export_mne_epochs`](https://eeglabio.readthedocs.io/en/latest/generated/eeglabio.utils.export_mne_epochs.html) from eeglabio. Do you think using [`export_set`](https://eeglabio.readthedocs.io/en/latest/generated/eeglabio.epochs.export_set.html#eeglabio-epochs-export-set) would make it easier? This way we can make mne and eeglabio being fully independent from each other

AlexLepauvre commented:
For the naming conventions, I called it modes in reference to the original function from spm, where they speak about the spatial mode of the data, because they typically do this type of analysis after some sort of dimensionality reduction (if I understand properly). But yeah, agree that here it might be more intuitive to call it n_channels, we can specify in the doc that it can be whatever features users fancy. 

As for using eeglabio, agree that export_set is better. I will adjust.

I will make a push with all of that as soon as I have figured out the effect size computation.

AlexLepauvre committed:
Add effect size specification

Specify distance between the two classes using the mahalanobis distance. This is used to calculate a factor to multiply the betas by, essentially scaling up or down the prominence of the injected pattern. This is nicely validated in simulation, where we can see a correspondance between the effect size and the decoding accuracy

AlexLepauvre committed:
Add kernel normalization

AlexLepauvre committed:
Update notebooks

AlexLepauvre committed:
Replace n_modes with n_channels

Change name of variable to match MNE conventions more closely

AlexLepauvre committed:
Use export set instead of export_mne_epochs

Makes eeglab export independent from MNE

qian-chu commented:
Wonderful! A few more suggestions (I can push on my end if you agree):

- Since we now say "channel", does it make sense to call `spat_cov` `ch_cov` instead?
- What is the scenario that users would input a customized `X` in `export_to_mne` and `export_to_eeglab`? Seems weird to me that people would want to change the design matrix after simulating them

AlexLepauvre commented:
Will change spat_cov to ch_cov to make it more explicit.
For the specification of the separate X, it's for interactions. The simulation scripts expect fully specified design matrix, one column per effect. But for exporting to mne or eeglab, that's redundant. In my examples, I have category: face, object, attention: attended, unattended. And then I have a third column: category x attention: face & attended, face&unattended... It would be really redundant to have:
trial 1: face/attended/face&attended...

Does that make sense?

qian-chu commented:
> Will change spat_cov to ch_cov to make it more explicit. For the specification of the separate X, it's for interactions. The simulation scripts expect fully specified design matrix, one column per effect. But for exporting to mne or eeglab, that's redundant. In my examples, I have category: face, object, attention: attended, unattended. And then I have a third column: category x attention: face & attended, face&unattended... It would be really redundant to have: trial 1: face/attended/face&attended...
> 
> Does that make sense?

Sorry I missed that part! Should have read the tutorial in more detail. Yes that makes sense to me

qian-chu committed:
rename spat_cov to ch_cov

AlexLepauvre committed:
BF: variable effect size for each subject

When specifying inter subjects variability, effect amplitude can be negative, requiring adjustment to the kernel scaling

AlexLepauvre committed:
Update meeg_simulator_intersubject_noise.ipynb

AlexLepauvre committed:
Create meeg_simulator_effect_size.ipynb

Add notebook describing how to play around with effect sizes.

AlexLepauvre committed:
BF: typo

AlexLepauvre committed:
Update paper

To add all the latest details

AlexLepauvre commented:
Alright, I think that's it for my change. Are we good to merge? 

In the feature, we should avoid working two at the same time on the notebooks. I was still doing some minor changes on my end, but since they can't easily be tracked with git, conflict resolution is not possible

AlexLepauvre committed:
Update paper.md

Slights improvements and add authors

qian-chu commented:
Sorry about that, I modified the notebooks when replacing all spat_cov. Currently it looks good to me! Let's merge to mark the end of classifying the package. I would love to contribute to other parts too (tutorial, paper etc.)

qian-chu committed:
Update paper.md

qian-chu committed:
add paper metainfo and paper compilation check

github-actions[bot] committed:
Format code with ruff

AlexLepauvre committed:
Update notebook for effect sizes

AlexLepauvre committed:
Update notebook for effect sizes

Bug in the json, fixed now

AlexLepauvre committed:
BF: replace bm with boldsymbol

bm requires an additional package

github-actions[bot] committed:
Format code with ruff

AlexLepauvre committed:
Update paper.md

AlexLepauvre committed:
Merge branch 'class' of https://github.com/qian-chu/meeg_simulator into pr/1

AlexLepauvre committed:
Update paper.md

Add a second dollar sign to prevent compilation issue

AlexLepauvre committed:
Update draft-pdf.yml

Commit paper to repo. By default, it is an artifact from the action

AlexLepauvre committed:
(auto) Paper PDF Draft

AlexLepauvre committed:
Update paper.md

AlexLepauvre committed:
(auto) Paper PDF Draft

AlexLepauvre committed:
Update toolbox name

Replacing meeg_simulator with multisim. We will need to rename the repo, but we will do that in upstream

AlexLepauvre committed:
Merge branch 'class' of https://github.com/qian-chu/meeg_simulator into pr/1

AlexLepauvre committed:
(auto) Paper PDF Draft

AlexLepauvre commented:
All good for the merge :)
